experimental/include/experimental/Support/FtdSupport.h

```diff
@@ -7,41 +7,92 @@
 //===----------------------------------------------------------------------===//
 //
 // Declares some utility functions which are useful for both the fast token
-// delivery algorithm and for the GSA anlaysis pass. All the functions are about
-// anlayzing relationships between blocks and handshake operations.
+// delivery algorithm and for the GSA analysis pass. All the functions are about
+// analyzing relationships between blocks and handshake operations.
 //
 //===----------------------------------------------------------------------===//
 
-#ifndef DYNAMATIC_SUPPORT_FTD_SUPPORT_H
-#define DYNAMATIC_SUPPORT_FTD_SUPPORT_H
+#ifndef DYNAMATIC_HANDSHAKE_SUPPORT_SUPPORT_H
+#define DYNAMATIC_HANDSHAKE_SUPPORT_SUPPORT_H
 
-#include "dynamatic/Support/LLVM.h"
+#include "dynamatic/Dialect/Handshake/HandshakeOps.h"
 #include "experimental/Support/BooleanLogic/BoolExpression.h"
 #include "mlir/Analysis/CFGLoopInfo.h"
 
 namespace dynamatic {
 namespace experimental {
 namespace ftd {
 
-/// Gets all the paths from block "start" to block "end" using a dfs search. If
-/// "blockToTraverse" is non null, then we want the paths having that block in
-/// the path; if "blocksToAvoid" is non empty, then we want the paths which do
-/// not cross those paths.
+/// Class to associate an index to each block, so that if block Bi dominates
+/// block Bj then i < j. While this is guaranteed by the MLIR CFG construction,
+/// it cannot really be given for granted, thus it is more convenient to have a
+/// custom indexing.
+class BlockIndexing {
+
+  /// Map to store the connection between indexes and blocks.
+  DenseMap<unsigned, Block *> indexToBlock;
+
+  /// Map to store the connection between blocks and indexes.
+  DenseMap<Block *, unsigned> blockToIndex;
+
+public:
+  /// Build the map out of a region.
+  BlockIndexing(mlir::Region &region);
+
+  /// Get a block out of an index.
+  Block *getBlockFromIndex(unsigned index) const;
+
+  /// Get a block out of a string condition in the format `cX` where X is a
+  /// number.
+  Block *getBlockFromCondition(const std::string &condition) const;
+
+  /// Get the index of a block.
+  unsigned getIndexFromBlock(Block *bb) const;
+
+  /// Return true if the index of bb1 is greater than then index of bb2.
+  bool greaterIndex(Block *bb1, Block *bb2) const;
```

Comment:
I would find a name like `isGreater` a bit clearer. Generally for function which return bool I like to name them so that they can be read out in plain old English when used in conditions like

```cpp
if (isGreater(bb1, bb2)) {}
```

Same comment for `lessIndex`.